File: graphs/4/shortest_paths/shortest_paths.cpp
#include <iostream>
#include <limits>
#include <vector>
#include <queue>
#include <string>

using std::vector;
using std::queue;
using std::pair;
using std::priority_queue;
long long inf = std::numeric_limits<long long>::max();


int BellmanFord(vector<vector<int> >&adj, vector<vector<int> > &cost, int s, vector<long long> &distance, 
                vector<bool> &reachable, vector<bool> &shortest, bool is_second = 0) {
  int n = adj.size();
  for (int k = 0; k < n - 1; k++) {
    bool f = 0;
    for (int i = 0; i < n; i++) {
      if (!reachable[i] || distance[i] >= inf) continue;
      for (int j = 0; j < adj[i].size(); j++) {
        int c = adj[i][j];
        if (!reachable[c]) continue;
        long long relax = distance[i] + cost[i][j];
        if (distance[c] > relax) {
          distance[c] = relax;
          f = 1;
          if (is_second) return c;
          shortest[c] = 0;
        } 
      }
    }
    if (!f) break;
  }
  return -1;
}


void dfs(vector<vector<int> > &adj, int x, vector<bool>&rchbl, bool filler, vector<bool> &u) {
  u[x] = 1;
  rchbl[x] = filler;
  for (int i = 0; i < adj[x].size(); i++) {
    int next = adj[x][i];
    if (!u[next]) {
      dfs(adj, next, rchbl, filler, u);
    }
  }
} 

void shortest_paths(vector<vector<int> > &adj, vector<vector<int> > &cost, int s, vector<long long> &distance, vector<bool> &reachable, vector<bool> &shortest) {
  //write your code here
  int n = adj.size();
  vector<bool> u(n, 0);
  dfs(adj, s, reachable, 1, u);
  BellmanFord(adj, cost, s, distance, reachable, shortest, 0);
  for (int i = 0; i < shortest.size(); i++) shortest[i] = 1;
  int cycle_node = BellmanFord(adj, cost, s, distance, reachable, shortest, 1);
  for (int i = 0; i < u.size(); i++) u[i] = 0;
  if (cycle_node != -1) dfs(adj, cycle_node, shortest, 0, u);
}

int main() {
  // std::string file_path = __FILE__;
  // #if defined(__unix__) || defined(__APPLE__)
  //   std::string test_path = file_path.substr(0, file_path.rfind("/")) + "/test/02";
  // #else
  //   std::string test_path = file_path.substr(0, file_path.rfind("\\")) + "/test/02";
  // #endif 
  // freopen(test_path.c_str(), "r", stdin);


  int n, m, s;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  vector<vector<int> > cost(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y, w;
    std::cin >> x >> y >> w;
    adj[x - 1].push_back(y - 1);
    cost[x - 1].push_back(w);
  }
  std::cin >> s;
  s--;
  vector<long long> distance(n, inf);
  distance[s] = 0;
  vector<bool> reachable(n, 0);
  vector<bool> shortest(n, 1);
  shortest_paths(adj, cost, s, distance, reachable, shortest);
  for (int i = 0; i < n; i++) {
    if (!reachable[i]) {
      std::cout << "*\n";
    } else if (!shortest[i]) {
      std::cout << "-\n";
    } else {
      std::cout << distance[i] << "\n";
    }
  }
}

File: graphs/4/shortest_paths/shortest_paths.py
#Uses python3

import sys, os
import queue
inf = 10**19

#WA 19
def BellmanFord(adj, cost, d, shortest, rchbl, is_second = False):
    n = len(adj)
    for _ in range(n - 1):
        f = False
        for i in range(n):
            for j in range(len(adj[i])):
                c = adj[i][j]
                if not rchbl[c]: continue
                if d[c] > d[i] + cost[i][j]:
                    if is_second: return c
                    d[c] = d[i] + cost[i][j]
                    f = True
                    shortest[c] = 0
        if not f:
            break
    return -1
        
def dfs(adj, s, u, rchbl, filler = 1):
    u[s] = True
    rchbl[s] = filler
    for i in adj[s]:
        if not u[i]:
            rchbl[i] = 1
            dfs(adj, i, u, reachable, filler)
        


def shortet_paths(adj, cost, s, d, reachable, shortest):
    #write your code here
    n = len(adj)
    u = [False] * n
    dfs(adj, s, u, reachable)
    BellmanFord(adj, cost, d, shortest, reachable)
    for i in range(n): shortest[i] = 1
    cycle_node = BellmanFord(adj, cost, d, shortest, reachable)
    for i in range(n): u[i] = False
    if not cycle_node == -1: dfs(adj, cycle_node, u, shortest, filler = 0)
    return shortest

if __name__ == '__main__':
    # sys.stdin = open(os.path.abspath(os.path.dirname(__file__)) + r'\test\01', 'r')
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(zip(data[0:(3 * m):3], data[1:(3 * m):3]), data[2:(3 * m):3]))
    data = data[3 * m:]
    adj = [[] for _ in range(n)]
    cost = [[] for _ in range(n)]
    for ((a, b), w) in edges:
        adj[a - 1].append(b - 1)
        cost[a - 1].append(w)
    s = data[0]
    s -= 1
    d = [inf] * n
    reachable = [0] * n
    shortest = [1] * n
    d[s] = 0
    shortet_paths(adj, cost, s, d, reachable, shortest)
    for x in range(n):
        if reachable[x] == 0:
            print('*')
        elif shortest[x] == 0:
            print('-')
        else:
            print(d[x])

